upstash_semantic_cache/semantic_cache.py

```diff
@@ -0,0 +1,183 @@
+        """
+        response = self._query_key(key)
+        if response is None or response.score <= self.min_proximity:
+            return None
+        return response.metadata["value"]
+
+    def lookup(
+        self, prompt: str, llm_string: Optional[str] = None
+    ) -> Optional[List[Generation]]:
+        """
+        Converts the JSON string to generations and returns them.
+        It is called by Langchain to check if the generations are in the cache.
+
+        Args:
+            prompt (str): The prompt to lookup in the cache.
+            llm_string (Optional[str]): Optional string for LLM context.
+
+        Returns:
+            Optional[List[Generation]]:
+            If response is found in the cache; otherwise, None.
+        """
+        result = self.get(prompt)
+        return self._loads_generations(result) if result else None
+
+    def update(
+        self,
+        prompt: str,
+        llm_string: Optional[str] = None,
+        result: Optional[str] = None,
+    ) -> None:
+        """
+        Converts the generations to a JSON string and stores it in the cache.
+        It is called by Langchain to upload the generations to the cache.
+
+        Args:
+            prompt (str): The prompt to update in the cache.
+            llm_string (Optional[str]): Optional string for LLM context.
+            result (Optional[str]): The result to store in the cache.
+        """
+        self.set(prompt, self._dumps_generations(result))
+
+    def set(self, key: Union[str, List[str]], value: Union[str, List[str]]) -> None:
+        if isinstance(key, list) and isinstance(value, list):
+            batch = [
+                (
+                    self._hash_key(self._namespaced_key(k)),
+                    self._namespaced_key(k),
+                    {"value": v},
+                )
+                for k, v in zip(key, value)
+            ]
+            self.index.upsert(batch)
+        else:
+            self.index.upsert(
+                [
+                    (
+                        self._hash_key(self._namespaced_key(key)),
+                        self._namespaced_key(key),
+                        {"value": value},
+                    )
+                ]
+            )
+
+    def delete(self, key: Union[str, List[str]]) -> None:
+        """
+        Deletes the key from the cache.
+
+        Args:
+            key (str): The key (or keys) to delete from the cache.
+        """
+        if isinstance(key, list):
+            batch = []
+            for current_key in key:
+                batch.append(self._hash_key(current_key))
+            self.index.delete(batch)
+        else:
+            self.index.delete([self._hash_key(key)])
+
+    def flush(self) -> None:
+        """
+        Resets the cache, removing all keys and values.
+        """
+        self.index.reset()
+
+    def _namespaced_key(self, key: Union[str, List[str]]) -> str:
+        return f"{self.namespace}:{key}" if self.namespace else key
```

Comment:
we should not do this, but pass the `self.namespace` to the `index` methods. So, if you do `self.index.delete([self._hash_key(key)])` now, you should just do `self.index.delete([self._hash_key(key)], namespace=self.namespace)`